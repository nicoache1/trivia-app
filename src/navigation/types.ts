import { ParamListBase } from '@react-navigation/native'
import { NativeStackScreenProps } from '@react-navigation/native-stack'
import React from 'react'

import { Routes } from './routes'

export interface ParamList extends ParamListBase {
  [Routes.Welcome]: undefined
  [Routes.ConfigureGame]: undefined
  [Routes.Game]: undefined
  [Routes.Results]: undefined
  [Routes.AnimatedSplash]: undefined
}

export type SceneProps<T extends keyof ParamList> = React.FC<
  NativeStackScreenProps<ParamList, T>
>

export type NavigationProp<T extends keyof ParamList> = NativeStackScreenProps<
  ParamList,
  T
>['navigation']
