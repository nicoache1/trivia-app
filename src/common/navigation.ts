import {
  NavigationContainerRef,
  ParamListBase,
  StackActions,
} from '@react-navigation/native'
import * as React from 'react'

import { ParamList } from '../navigation/types'

export const navigationRef: React.RefObject<NavigationContainerRef<ParamList>> =
  React.createRef()

export function navigate(name: string, params?: ParamListBase[string]) {
  navigationRef.current?.navigate(name, params)
}

export function push(name: string, params?: ParamListBase[string]) {
  navigationRef.current?.dispatch(StackActions.push(name, params))
}

export function popToTop() {
  navigationRef.current?.dispatch(StackActions.popToTop())
}

export function pop() {
  navigationRef.current?.dispatch(StackActions.pop())
}
