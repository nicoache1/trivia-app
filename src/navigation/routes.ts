export enum Routes {
  Game = 'Game',
  ConfigureGame = 'ConfigureGame',
  AnimatedSplash = 'AnimatedSplash',
  Results = 'Results',
  Welcome = 'Welcome',
}
