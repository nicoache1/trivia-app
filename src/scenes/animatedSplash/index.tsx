import SplashImage from 'assets/splashShape.png'
import * as SplashScreen from 'expo-splash-screen'
import React, { useLayoutEffect } from 'react'
import { Dimensions, StyleSheet, View } from 'react-native'
import Animated, {
  interpolate,
  runOnJS,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated'
import { Routes } from 'src/navigation/routes'
import { SceneProps } from 'src/navigation/types'
import { useTheme } from 'src/styles/Theme'

import { styles } from './styles'

const { width, height } = Dimensions.get('window')

export const AnimatedSplash: SceneProps<Routes.AnimatedSplash> = ({
  navigation,
}) => {
  useLayoutEffect(() => {
    navigation.setOptions({ headerShown: false })
  }, [navigation])

  const {
    Theme: { colors },
  } = useTheme()

  const progress = useSharedValue(0)
  const imageAnimation = useSharedValue(0)

  const onImageLoaded = async () => {
    await SplashScreen.hideAsync()
    progress.value = withTiming(1, { duration: 700 }, () => {
      imageAnimation.value = withTiming(1, {}, finished => {
        if (finished) {
          runOnJS(navigation.replace)(Routes.Welcome)
        }
      })
    })
  }

  const animatedStyle = useAnimatedStyle(() => ({
    backgroundColor: colors.SECONDARY,
    transform: [
      { translateY: width / 2 },
      { rotate: `${interpolate(progress.value, [0, 1], [-25, 0])}deg` },
      { translateX: interpolate(progress.value, [0, 1], [-height, 0]) },
      { translateY: -width / 2 },
    ],
    zIndex: 1,
  }))

  const animatedImage = useAnimatedStyle(() => ({
    height: interpolate(imageAnimation.value, [0, 0.5, 1], [70, 80, 0]),
    opacity: interpolate(imageAnimation.value, [0, 1], [1, 0]),
    zIndex: 100,
  }))

  return (
    <View style={styles.container}>
      <Animated.Image
        onLoadEnd={onImageLoaded}
        source={SplashImage}
        style={animatedImage}
        resizeMode={'contain'}
      />
      <Animated.View
        style={[animatedStyle, { ...StyleSheet.absoluteFillObject }]}
      />
    </View>
  )
}
