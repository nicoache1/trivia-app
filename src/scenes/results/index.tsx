import { localization } from 'localization'
import React, { useLayoutEffect } from 'react'
import { Dimensions, FlatListProps, StyleSheet, View } from 'react-native'
import { FlatList } from 'react-native-gesture-handler'
import { Path, Svg } from 'react-native-svg'
import { useSelector } from 'react-redux'

import { SceneContainer } from '../../components/SceneContainer'
import { StyledButton } from '../../components/StyledButton'
import { ButtonMode } from '../../components/StyledButton/ButtonMode'
import { StyledContainer } from '../../components/StyledContainer'
import { Routes } from '../../navigation/routes'
import { SceneProps } from '../../navigation/types'
import { useAppDispatch } from '../../store'
import { restartGameAction } from '../../store/storage/game'
import { selectGameConfiguration } from '../../store/storage/game/selectors'
import { useTheme } from '../../styles/Theme'
import { QuestionResult } from '../../types/questions'
import { Header } from './Header'
import { ResultItem } from './ResultItem'
import { ITEM_HEIGHT } from './ResultItem/styles'
import { styles } from './styles'

const { width, height } = Dimensions.get('window')

const renderItem: FlatListProps<QuestionResult>['renderItem'] = ({ item }) => (
  <ResultItem item={item} />
)

const keyExtractor: FlatListProps<QuestionResult>['keyExtractor'] = ({ id }) =>
  id

const getItemLayout: FlatListProps<QuestionResult>['getItemLayout'] = (
  _,
  index,
) => ({
  index,
  length: ITEM_HEIGHT,
  offset: ITEM_HEIGHT * index,
})

export const Results: SceneProps<Routes.Results> = ({ navigation }) => {
  const {
    Theme: { colors },
  } = useTheme()

  useLayoutEffect(() => {
    navigation.setOptions({
      headerShown: false,
    })
  }, [colors.WHITE, navigation])

  const dispatch = useAppDispatch()

  const game = useSelector(selectGameConfiguration)

  const onPressRestartGame = () => {
    dispatch(restartGameAction())
    navigation.reset({ index: 0, routes: [{ name: Routes.ConfigureGame }] })
  }

  const total = game.questionAmount.valueOf()
  const correctOnes = game.results.filter(item => item.correct).length

  return (
    <SceneContainer
      barStyle={'light-content'}
      style={[styles.container, { backgroundColor: colors.SECONDARY }]}>
      <View style={{ ...StyleSheet.absoluteFillObject }}>
        <Svg width={width} height={height} viewBox="0 0 375 812" fill="none">
          <Path
            d="M-382.5 612.5L-835 243.5V948.5H834V264.5L487 673L39.5 -76L-382.5 612.5Z"
            fill="#142664"
            fillOpacity="0.4"
          />
        </Svg>
      </View>
      <FlatList
        ListHeaderComponent={<Header total={total} correctOnes={correctOnes} />}
        data={game.results}
        renderItem={renderItem}
        keyExtractor={keyExtractor}
        getItemLayout={getItemLayout}
      />
      <StyledContainer style={styles.footer}>
        <StyledButton
          mode={ButtonMode.DARK}
          label={localization('restartGameLabel')}
          onPress={onPressRestartGame}
        />
      </StyledContainer>
    </SceneContainer>
  )
}
