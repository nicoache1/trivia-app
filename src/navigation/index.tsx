import { createNativeStackNavigator } from '@react-navigation/native-stack'
import React from 'react'

import { AnimatedSplash } from '../scenes/animatedSplash'
import { ConfigureGame } from '../scenes/configureGame'
import { Game } from '../scenes/game'
import { Results } from '../scenes/results'
import { Welcome } from '../scenes/welcome'
import { Routes } from './routes'
import { ParamList } from './types'

const Stack = createNativeStackNavigator<ParamList>()

export const AppNavigator: React.FC<{}> = ({}) => {
  return (
    <Stack.Navigator
      screenOptions={{
        contentStyle: {
          backgroundColor: 'white',
        },
      }}>
      <Stack.Screen name={Routes.AnimatedSplash} component={AnimatedSplash} />
      <Stack.Screen
        name={Routes.Welcome}
        component={Welcome}
        options={{ presentation: 'modal' }}
      />
      <Stack.Screen name={Routes.ConfigureGame} component={ConfigureGame} />
      <Stack.Screen name={Routes.Game} component={Game} />
      <Stack.Screen name={Routes.Results} component={Results} />
    </Stack.Navigator>
  )
}
