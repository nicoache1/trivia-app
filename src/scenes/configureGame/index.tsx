import ConfigureImage from 'assets/configure.png'
import { localization } from 'localization'
import React, { useLayoutEffect } from 'react'
import { Controller, useForm } from 'react-hook-form'
import { Dimensions, Image, View } from 'react-native'
import { Background } from 'src/components/Background'
import { SceneContainer } from 'src/components/SceneContainer'
import { StyledButton } from 'src/components/StyledButton'
import { ButtonMode } from 'src/components/StyledButton/ButtonMode'
import { StyledContainer } from 'src/components/StyledContainer'
import { StyledText } from 'src/components/StyledText'
import { Routes } from 'src/navigation/routes'
import { SceneProps } from 'src/navigation/types'
import { useAppDispatch } from 'src/store'
import { setGameAction } from 'src/store/storage/game'
import { useTheme } from 'src/styles/Theme'
import { AppTypography, ColorPalette } from 'src/styles/types'
import { Game, getDifficulties, getQuestionAmounts } from 'src/types/game'

import { renderInputControl, renderSegmentedControl } from './formRenders'
import { styles } from './styles'
import { ConfigureGameForm } from './types'

const { width, height } = Dimensions.get('window')

export const ConfigureGame: SceneProps<Routes.ConfigureGame> = ({
  navigation,
}) => {
  const {
    Theme: { colors },
  } = useTheme()

  const dispatch = useAppDispatch()

  useLayoutEffect(() => {
    navigation.setOptions({
      headerShown: false,
    })
  }, [navigation])

  const { control, handleSubmit } = useForm<ConfigureGameForm>({
    defaultValues: {
      amount: 0,
      difficulty: 0,
      name: localization('defaultName'),
    },
  })

  const difficulties = getDifficulties()
  const questions = getQuestionAmounts()

  const onSubmit = (data: ConfigureGameForm) => {
    const newGame: Game = {
      difficulty: difficulties[data.difficulty],
      player: data.name,
      questionAmount: questions[data.amount],
      results: [],
    }
    dispatch(setGameAction(newGame))
    navigation.navigate(Routes.Game)
  }

  return (
    <View
      style={[styles.screenContainer, { backgroundColor: colors.SECONDARY }]}>
      <SceneContainer style={styles.scene} barStyle={'light-content'}>
        <View style={styles.headerContainer}>
          <Background />
          <Image
            source={ConfigureImage}
            style={{ height: height / 2.5, width: width * 0.9 }}
            resizeMode={'contain'}
          />
          <View style={styles.textContainer}>
            <StyledText
              color={ColorPalette.TERTIARY}
              typography={AppTypography.H1}>
              {localization('configureTitle1')}
            </StyledText>
            <StyledText
              color={ColorPalette.WHITE}
              typography={AppTypography.H1}>
              {localization('configureTitle2')}
            </StyledText>
          </View>
        </View>
        <View style={styles.formContainer}>
          <Controller
            control={control}
            rules={{
              required: true,
            }}
            render={renderInputControl}
            name={'name'}
          />
          <StyledContainer>
            <View style={styles.subtitle}>
              <StyledText
                color={ColorPalette.WHITE}
                typography={AppTypography.P1_BOLD}>
                {localization('difficultyLabel')}
              </StyledText>
            </View>
            <Controller
              control={control}
              rules={{
                required: true,
              }}
              render={renderSegmentedControl<'difficulty'>(
                difficulties,
                colors.WHITE,
              )}
              name={'difficulty'}
            />
          </StyledContainer>
          <StyledContainer>
            <View style={styles.subtitle}>
              <StyledText
                color={ColorPalette.WHITE}
                typography={AppTypography.P1_BOLD}>
                {localization('questionAmountLabel')}
              </StyledText>
            </View>
            <Controller
              control={control}
              rules={{
                required: true,
              }}
              render={renderSegmentedControl<'amount'>(
                questions.map(item => `${item}`),
                colors.WHITE,
              )}
              name={'amount'}
            />
          </StyledContainer>
        </View>
        <StyledContainer style={styles.buttonContainer}>
          <StyledButton
            mode={ButtonMode.DARK}
            label={localization('getStartedLabel')}
            onPress={handleSubmit(onSubmit)}
          />
        </StyledContainer>
      </SceneContainer>
    </View>
  )
}
