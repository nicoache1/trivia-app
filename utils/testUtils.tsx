import type { PreloadedState } from '@reduxjs/toolkit'
import type { RenderOptions } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { AppStore, RootState, setupStore } from 'src/store'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export function renderWithProvider(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
  }: ExtendedRenderOptions = {},
) {
  return <Provider store={store}>{ui}</Provider>
}
