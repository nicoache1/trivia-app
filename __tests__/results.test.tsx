import { render } from '@testing-library/react-native'
import React from 'react'
import { Routes } from 'src/navigation/routes'
import { Results } from 'src/scenes/results'

import {
  createNavigationPropMock,
  createRoutePropMock,
} from '../__mocks__/navigationMocks'
import { renderWithProvider } from '../utils/testUtils'

test('Expect Results to be rendered correctly with default', async () => {
  const mockRoute = createRoutePropMock<Routes.Results>(Routes.Results)

  const mockNavigation = createNavigationPropMock<Routes.Results>()

  const enhancedComponent = renderWithProvider(
    <Results route={mockRoute} navigation={mockNavigation} />,
  )

  const renderedSnapshot = render(enhancedComponent).toJSON()

  expect(renderedSnapshot).toMatchSnapshot()
})
