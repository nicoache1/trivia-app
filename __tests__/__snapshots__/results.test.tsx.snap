// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect Results to be rendered correctly with default 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
    ]
  }
  style={
    Array [
      Object {
        "flexGrow": 1,
      },
      Object {
        "backgroundColor": "#FFFFFF",
      },
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "backgroundColor": "rgb(108,92,231)",
        },
      ],
    ]
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight={1334}
      bbWidth={750}
      fill="none"
      focusable={false}
      height={1334}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "flex": 0,
            "height": 1334,
            "width": 750,
          },
        ]
      }
      vbHeight={812}
      vbWidth={375}
      width={750}
    >
      <RNSVGGroup
        fill={null}
        propList={
          Array [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M-382.5 612.5L-835 243.5V948.5H834V264.5L487 673L39.5 -76L-382.5 612.5Z"
          fill={4279510628}
          fillOpacity={0.4}
          propList={
            Array [
              "fill",
              "fillOpacity",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <RCTScrollView
    ListHeaderComponent={
      <Header
        correctOnes={0}
        total={5}
      />
    }
    collapsable={false}
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    getItemLayout={[Function]}
    handlerTag={1}
    handlerType="NativeViewGestureHandler"
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    renderScrollComponent={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "marginTop": 60,
              "paddingHorizontal": 20,
              "paddingVertical": 10,
            }
          }
        >
          <Text
            accessible={true}
            color="WHITE"
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 32,
                "fontStyle": "normal",
                "fontWeight": "700",
              }
            }
            typography="H1"
          >
            resultsSectionTitle
          </Text>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "transparent",
              "height": 400,
              "paddingHorizontal": 20,
              "paddingVertical": 10,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          >
            <RNSVGSvgView
              align="xMidYMid"
              bbHeight={300}
              bbWidth={300}
              fill="none"
              focusable={false}
              height={300}
              meetOrSlice={0}
              minX={0}
              minY={0}
              style={
                Array [
                  Object {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  Object {
                    "flex": 0,
                    "height": 300,
                    "width": 300,
                  },
                ]
              }
              vbHeight={300}
              vbWidth={300}
              width={300}
            >
              <RNSVGGroup
                fill={null}
                propList={
                  Array [
                    "fill",
                  ]
                }
              >
                <RNSVGCircle
                  cx={150}
                  cy={150}
                  fill={0}
                  propList={
                    Array [
                      "fill",
                      "stroke",
                      "strokeWidth",
                    ]
                  }
                  r={135}
                  stroke={1711323283}
                  strokeWidth={30}
                />
                <RNSVGCircle
                  cx={150}
                  cy={150}
                  fill={0}
                  propList={
                    Array [
                      "fill",
                      "stroke",
                      "strokeWidth",
                      "strokeDasharray",
                      "strokeDashoffset",
                    ]
                  }
                  r={135}
                  stroke={4278237332}
                  strokeDasharray={
                    Array [
                      "848.2300164692441",
                      "848.2300164692441",
                    ]
                  }
                  strokeDashoffset={848.2300164692441}
                  strokeWidth={30}
                />
              </RNSVGGroup>
            </RNSVGSvgView>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "position": "absolute",
                }
              }
            >
              <Text
                accessible={true}
                color="WHITE"
                style={
                  Object {
                    "color": "#FFFFFF",
                    "fontSize": 32,
                    "fontStyle": "normal",
                    "fontWeight": "700",
                    "textAlign": "center",
                  }
                }
                typography="H1"
              >
                0
              </Text>
              <Text
                accessible={true}
                color="WHITE"
                style={
                  Object {
                    "color": "#FFFFFF",
                    "fontSize": 32,
                    "fontStyle": "normal",
                    "fontWeight": "700",
                    "textAlign": "center",
                  }
                }
                typography="H1"
              >
                of 5
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "transparent",
              "paddingHorizontal": 20,
              "paddingVertical": 10,
            }
          }
        >
          <Text
            accessible={true}
            color="WHITE"
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 32,
                "fontStyle": "normal",
                "fontWeight": "700",
              }
            }
            typography="H1"
          >
            detailsSectionTitle
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "height": 100,
        "paddingHorizontal": 20,
        "paddingVertical": 10,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderRadius": 10,
            "height": 56,
            "justifyContent": "center",
            "paddingVertical": 10,
            "width": "100%",
          },
          Object {
            "opacity": 1,
          },
        ]
      }
    >
      <View
        animatedStyle={
          Object {
            "value": Object {},
          }
        }
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "rgb(0,184,148)",
            "borderRadius": 10,
            "height": 56,
            "justifyContent": "center",
            "paddingVertical": 10,
            "width": "100%",
          }
        }
      >
        <Text
          accessible={true}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontSize": 16,
                "fontStyle": "normal",
                "fontWeight": "700",
              },
              Object {
                "alignItems": "center",
              },
              Object {
                "color": "#FFFFFF",
              },
            ]
          }
          typography="P1_BOLD"
        >
          restartGameLabel
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
